pkg/gadget-collection/gadgets/top/ebpf/gadget.go

```diff
@@ -0,0 +1,273 @@
+ - %s: Maximum rows to print. (default %d)
+ - %s: The field to sort the results by (%s). (default %s)`
+	return fmt.Sprintf(t, types.IntervalParam, types.IntervalDefault,
+		types.MaxRowsParam, types.MaxRowsDefault,
+		types.SortByParam, strings.Join(types.SortBySlice, ","), types.SortByDefault)
+}
+
+func (f *TraceFactory) OutputModesSupported() map[string]struct{} {
+	return map[string]struct{}{
+		"Stream": {},
+	}
+}
+
+func deleteTrace(name string, t interface{}) {
+	trace := t.(*Trace)
+	if trace.started {
+		trace.stop()
+	}
+}
+
+func (f *TraceFactory) Operations() map[string]gadgets.TraceOperation {
+	n := func() interface{} {
+		return &Trace{
+			resolver: f.Resolver,
+		}
+	}
+
+	return map[string]gadgets.TraceOperation{
+		"start": {
+			Doc: "Start ebpftop gadget",
+			Operation: func(name string, trace *gadgetv1alpha1.Trace) {
+				f.LookupOrCreate(name, n).(*Trace).Start(trace)
+			},
+		},
+		"stop": {
+			Doc: "Stop ebpftop gadget",
+			Operation: func(name string, trace *gadgetv1alpha1.Trace) {
+				f.LookupOrCreate(name, n).(*Trace).Stop(trace)
+			},
+		},
+	}
+}
+
+func (t *Trace) Start(trace *gadgetv1alpha1.Trace) {
+	if t.started {
+		trace.Status.State = "Started"
+		return
+	}
+
+	t.donechan = make(chan bool)
+	traceName := gadgets.TraceName(trace.ObjectMeta.Namespace, trace.ObjectMeta.Name)
+
+	// Enable stats collection
+	err := bpfstats.EnableBPFStats()
+	if err != nil {
+		trace.Status.OperationError = err.Error()
+		return
+	}
+
+	useFallbackIterator := false
+
+	// To resolve pids, we will first try to iterate using a bpf
+	// program. If that doesn't work, we will fall back to scanning
+	// all used fds in all processes /proc/$pid/fdinfo/$fd.
+	iter, err := piditer.NewPidIter()
+	if err != nil {
+		useFallbackIterator = true
+	}
+
+	t.started = true
+
+	go func() {
+		defer iter.Close()
+
+		var err error
+		var prog *ebpf.Program
+		var pids []*piditer.PidIterEntry
+		curID := ebpf.ProgramID(0)
+		nextID := ebpf.ProgramID(0)
+		timer := time.NewTicker(time.Second)
+		for {
+			select {
+			case <-t.donechan:
+				return
+			case <-timer.C:
+				curID = 0
+
+				stats := make([]types.Stats, 0)
+
+				for {
+					nextID, err = ebpf.ProgramGetNextID(curID)
+					if errors.Is(err, os.ErrNotExist) {
+						break
+					}
+					if nextID <= curID {
+						break
+					}
+					curID = nextID
+					prog, err = ebpf.NewProgramFromID(curID)
+					if err != nil {
+						continue
+					}
+					pi, err := prog.Info()
+					if err != nil {
+						continue
+					}
+					rt, _ := pi.Runtime()
+					rc, _ := pi.RunCount()
+
+					stats = append(stats, types.Stats{
+						ProgramID: uint32(curID),
+						Pids:      nil,
+						Name:      pi.Name,
+						Type:      pi.Type.String(),
+						Runtime:   int64(rt),
+						RunCount:  rc,
+					})
+
+					prog.Close()
+				}
+
+				pidmap := make(map[uint32][]*types.PidInfo)
+
+				if !useFallbackIterator {
+					pids, err = iter.DumpPids()
```

Comment:
It seems `var pids []*piditer.PidIterEntry` could be declared just before here since it's not reused at the next tick.